src/merkle_proof/poseidon_branch.rs

```diff
@@ -111,6 +116,41 @@ impl PoseidonBranch {
             levels: Vec::with_capacity(n),
         }
     }
+
+    /// Extends the branch to the specified length
+    pub fn extend(&mut self, target_depth: usize) -> usize {
+        let n_extensions = target_depth - self.levels.len();
+        while self.levels.len() < target_depth {
+            let old_root = mem::take(&mut self.root);
+            let flag = Scalar::from(0b1000);
+            let mut leaves = [Scalar::zero(); ARITY + 1];
+
+            leaves[0] = flag;
+            leaves[1] = old_root;
+
+            let level = PoseidonLevel { leaves, offset: 1 };
+            self.root = hash::merkle_level_hash_without_bitflags(&level);
+            self.levels.push(level);
+        }
+        n_extensions
+    }
+
+    /// Validate that all the hashes in the branch match up
+    pub fn valid(&self) -> bool {
+        let mut expected = None;
+
+        for level in &self.levels {
+            if let Some(expected) = expected {
+                let should_equal = level.leaves[level.offset];
+
+                if expected != should_equal {
+                    return false;
+                }
+            }
+            expected = Some(hash::merkle_level_hash_without_bitflags(level));
+        }
+        Some(self.root) == expected
+    }
```

Comment:
Looks like a duplication of the scalar implementation that we have for the opening. Where we do exactly the same.

Reply:
I did not see that, but i think redundancy is not bad in testing code

Reply:
Removed the redundant code